architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+Resources are the unit of exchange in the Radius API. All of our user-experiences like the CLI, Dashboard, and Bicep authoring are oriented around the catalog of resources. Because Radius is strongly-typed, the set of resource types and their schemas are also defined as a primary concept in Radius. In order for Radius to support each additional kind of resources, for example a PostgreSQL database, we must first have a type definition.
+
+The registration APIS for resource types determine how users add, change, or remove types from Radius.
+
+## Terms and definitions
+
+*Please read the [Radius API](https://docs.radapp.io/concepts/technical/api/) conceptual documentation. This document will heavily use the terminology and concepts defined there.*
+
+*Please see additional terms and context in the [overall user-defined-types](https://github.com/radius-project/design-notes/pull/56/files) design note.*
+
+| Term               | Definition                                                                                                                                                                                                                                                                                   |
+| ------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Resource Namespace | A group of related or similar resources that share a resource provider. eg: In `Applications.Core/containers` the resource namespace is `Applications.Core`.                                                                                                                                 |
+| Resource Provider  | A resource namespace and implementation of the resource namespace. Resource providers are a new resource type introduced by this proposal.                                                                                                                                                   |
+| Resource Type      | An API entity with CRUDL operations. Resource Types are part of a resource provider. Resource types are a new resource type introduced by this proposal.                                                                                                                                     |
+| OpenAPI/Schema     | A format for documenting HTTP-based APIs. We use OpenAPI v2 currently in Radius for client-sdk generation and validation of inbound requests. OpenAPI v3 is more widely adopted. In this document we're primarily concerned with the parts of OpenAPI pertaining to request/response bodies. |
```

Comment:
How much work would it be to move to OpenAPI v3? what benefits would we/users garner from it?

Reply:
This is probably most relevant to one of the upcoming docs, but I'll answer anyway 👍. 

It would be a small amount of work for us to move because we're using TypeSpec to abstract this away and to generate the OpenAPIv2 doc. Generating OpenAPIv3 would be a different command line parameter and possibly some additional annotations in TypeSpec. We could possibly generate both and use them for different purposes. The Azure/ARM toolchain for generating clients and validating RP APIs is based on v2.

The benefit for us is that we're not maintaining an old thing that the world has moved on from. The benefit for users is that they are much more likely to be familiar with v3. v3 has a bunch of minor design improvement compared to v2.